### PR TITLE
add OSPlatform.macOS, switch to OrdinalIgnoreCase for OSPlatform comparisons

### PR DESCRIPTION
**Pull request by adamsitnik**
I tried to introduce as small overhead as possible.

On Windows, there is basically no difference. On OSX there is no difference for "OSX" (old check), but the "macOS" check takes twice the time because we need two comparisons.

I've added the test cases that were missing in #39005 (ignore case comparison was not enabled then)

## Windows

```cs
private readonly OSPlatform _notCurrent = OSPlatform.Create("NOPE");
private readonly OSPlatform _current = OSPlatform.Windows;

[Benchmark]
public bool IsOSPlatform_True() => RuntimeInformation.IsOSPlatform(_current);

[Benchmark]
public bool IsOSPlatform_False() => RuntimeInformation.IsOSPlatform(_notCurrent);
```

```ini
BenchmarkDotNet=v0.12.1, OS=Windows 10.0.18363.900 (1909/November2018Update/19H2)
Intel Xeon CPU E5-1650 v4 3.60GHz, 1 CPU, 12 logical and 6 physical cores
```

Before:

|             Method |     Mean |     Error |    StdDev |   Median |      Min |      Max | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------- |---------:|----------:|----------:|---------:|---------:|---------:|------:|------:|------:|----------:|
|  IsOSPlatform_True | 3.160 ns | 0.0116 ns | 0.0103 ns | 3.158 ns | 3.145 ns | 3.183 ns |     - |     - |     - |         - |
| IsOSPlatform_False | 4.592 ns | 0.0139 ns | 0.0130 ns | 4.590 ns | 4.571 ns | 4.613 ns |     - |     - |     - |         - |

After:

|             Method |     Mean |     Error |    StdDev |   Median |      Min |      Max | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------- |---------:|----------:|----------:|---------:|---------:|---------:|------:|------:|------:|----------:|
|  IsOSPlatform_True | 3.177 ns | 0.0210 ns | 0.0197 ns | 3.171 ns | 3.152 ns | 3.213 ns |     - |     - |     - |         - |
| IsOSPlatform_False | 4.739 ns | 0.0126 ns | 0.0112 ns | 4.741 ns | 4.722 ns | 4.761 ns |     - |     - |     - |         - |

### OSX

```ini
BenchmarkDotNet=v0.12.1, OS=macOS Mojave 10.14.5 (18F132) [Darwin 18.6.0]
Intel Core i7-5557U CPU 3.10GHz (Broadwell), 1 CPU, 4 logical and 2 physical cores
```

```cs
private readonly OSPlatform _notCurrent = OSPlatform.Create("NOPE");
private readonly OSPlatform _current = OSPlatform.OSX;
private readonly OSPlatform _macOS = OSPlatform.macOS;

[Benchmark]
public bool IsOSPlatform_True() => RuntimeInformation.IsOSPlatform(_current);

[Benchmark]
public bool IsOSPlatform_macOS() => RuntimeInformation.IsOSPlatform(_macOS );

[Benchmark]
public bool IsOSPlatform_False() => RuntimeInformation.IsOSPlatform(_notCurrent);
```

Before:

|             Method |     Mean |     Error |    StdDev |   Median |      Min |      Max | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------- |---------:|----------:|----------:|---------:|---------:|---------:|------:|------:|------:|----------:|
|  IsOSPlatform_True | 2.901 ns | 0.0071 ns | 0.0063 ns | 2.901 ns | 2.892 ns | 2.915 ns |     - |     - |     - |         - |
| IsOSPlatform_macOS | 4.091 ns | 0.0092 ns | 0.0077 ns | 4.089 ns | 4.081 ns | 4.106 ns |     - |     - |     - |         - |
| IsOSPlatform_False | 4.415 ns | 0.0209 ns | 0.0163 ns | 4.411 ns | 4.402 ns | 4.463 ns |     - |     - |     - |         - |

After:

|             Method |     Mean |     Error |    StdDev |   Median |      Min |      Max | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------- |---------:|----------:|----------:|---------:|---------:|---------:|------:|------:|------:|----------:|
|  IsOSPlatform_True | 2.887 ns | 0.0057 ns | 0.0048 ns | 2.886 ns | 2.880 ns | 2.897 ns |     - |     - |     - |         - |
| IsOSPlatform_macOS | 7.972 ns | 0.0146 ns | 0.0130 ns | 7.968 ns | 7.956 ns | 7.996 ns |     - |     - |     - |         - |
| IsOSPlatform_False | 9.689 ns | 0.0378 ns | 0.0335 ns | 9.682 ns | 9.627 ns | 9.751 ns |     - |     - |     - |         - |

### CODE REVIEW

File: src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/RuntimeInformation.Unix.cs
Comment (jeffhandley) [inline src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/RuntimeInformation.Unix.cs:24]: Wow. 🥇 

```suggestion
            else if (name.Length == 3 && name[0] == 'O' && name[1] == 'S' && name[2] == 'X') // perf optimization; this is a hot path
```

**Review by jeffhandley — APPROVED** [review]
Looks good to me.  I'm going to add @jkotas as a reviewer too though; we should wait for his approval before merging.

File: src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/RuntimeInformation.Unix.cs
Comment (jkotas) [reply src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/RuntimeInformation.Unix.cs:24]: This optimization is pointless once you have paid for OrdinalIgnoreCase comparison in `osPlatform.Equals`

Comment (jkotas) [reply src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/RuntimeInformation.Unix.cs:24]: Also, I think it does not look right that the comparison for everything else is case-insensitive; but the OSX match is case sensitive.

File: src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/OSPlatform.cs
Comment (jkotas) [inline src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/OSPlatform.cs:58]:
```suggestion
            return obj is OSPlatform osPlatform && Equals(osPlatform);
```

File: src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/RuntimeInformation.Unix.cs
Comment (jkotas) [reply src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/RuntimeInformation.Unix.cs:24]: Ah, the OSX is coming from PAL, so we know that it will be all upper case.

If we want to make this API to be high-performance so that people do not ever have a reason to cache the result to avoid the performance overhead, I think we should cache the result of the "is current platform" check in the constructor.

Comment (jkotas) [reply src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/RuntimeInformation.Unix.cs:24]: I expect that the performance of the API is uneven due to case-insensitive comparison, e.g.:
- Checking for macOS on Linux is more expensive because both macOS and Linux have 5 characters
- Checking for Windows on Linux is cheaper because we will take early path out due to different string lengths

Caching the result in constructor would make the performance consistent.

Comment (jeffhandley) [reply src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/RuntimeInformation.Unix.cs:24]: > If we want to make this API to be high-performance so that people do not ever have a reason to cache the result to avoid the performance overhead, I think we should cache the result of the "is current platform" check in the constructor.

Is something like this what you're imagining, @jkotas?

**RuntimeInformation.Unix.cs** (platform-specific, to be defined on Windows and Browser as well)
``` c#
public static partial class RuntimeInformation
{
    internal static IsOSPlatformOrEquivalent(string osPlatform)
    {
        string name = s_osPlatformName ??= Interop.Sys.GetUnixName();

        if (osPlatform.Equals(name))
        {
            return true;
        }

        if (name.Equals("OSX"))
        {
            return osPlatform.Equals("MACOS", StringComparison.OrdinalIgnoreCase);
        }

        return false;
    }
}
```

**OSPlatform.cs**
``` c#
public readonly struct OSPlatform : IEquatable<OSPlatform>
{
    internal bool IsCurrentOSPlatform { get; set; }

    private OSPlatform(string osPlatform)
    {
        if (osPlatform == null) throw new ArgumentNullException(nameof(osPlatform));
        if (osPlatform.Length == 0) throw new ArgumentException(SR.Argument_EmptyValue, nameof(osPlatform));

        _osPlatform = osPlatform;
        IsCurrentOSPlatform = RuntimeInformation.IsOSPlatformOrEquivalent(_osPlatform);
    }
}
```

**RuntimeInformation.cs** (cross-platform)
``` c#
public static partial class RuntimeInformation
{
    public static bool IsOSPlatform(OSPlatform osPlatform)
    {
        return osPlatform.IsCurrentOSPlatform;
    }
}
```

Comment (jkotas) [reply src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/RuntimeInformation.Unix.cs:24]: Yes.

> platform-specific, to be defined on Windows and Browser as well

This should not be required to fix the throughput - the assembly can stay platform neutral for that. 

However, folding this tiny assembly into CoreLib and making it platform specific would be nice to reduce the startup costs.

Comment (jkotas) [reply src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/RuntimeInformation.Unix.cs:24]: > `internal bool IsCurrentOSPlatform { get; set; }`

This should not need to be `set`. Having internal readonly field instead of property would be fine too.

Comment (adamsitnik) [reply src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/RuntimeInformation.Unix.cs:24]: >  I think we should cache the result of the "is current platform" check in the constructor.

This approach also has some disadvantages:

* higher initialization cost - we need to perform this check for every known `OSPlatform` (10 as of today, most probably will grow over the time) when initializing [these](https://github.com/dotnet/runtime/blob/911e62bbe9558abd45bd67482e3634c1532219bc/src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/OSPlatform.cs#L12-L31) static fields
* won't make much difference for scenarios where consumer library is using `OSPlatform.Create` This is common for all libraries that target .NET Standard 2.0 and don't publish netcoreapp-specific packages.

To defend my proposal:

* the requirement of `OrdinalIgnoreCase` case comparison comes from #33331 and there are some good reasons why we need that. 
* For all the use cases except checking for `macOS` on `OSX`, there is no noticeable overhead (please see the numbers that I've posted for Windows in the PR description): 
     * the "hits" like checking `RuntimeInformation.IsOSPlatform(OSPlatform.Windows)` on Windows execute the perfect hot path (all letters match)
     * the "misses": the set of known inputs is limited. Most of the strings have different lengths, so the comparison returns quickly. Even if the length is the same, the first letter is different. Again, a quick return

If the new APIs were supposed to be as fast as possible, they should rather be based on `enum` rather than `string` and should definitely not allow for an input that is concatenated name and version. The usage that I see is targeting certain versions of `Windows`, `Android`, `iOS`, and `macOS` in netcoreapp-specific apps (and every version of .NET Core supports limited number of platforms which could be just part of enumeration). I am not sure if it has any usage on Linux (does the version refers to OS or kernel version?).

Comment (adamsitnik) [reply src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/RuntimeInformation.Unix.cs:24]: > However, folding this tiny assembly into CoreLib and making it platform specific would be nice to reduce the startup costs.

I really like this idea. Having it combined with an introduction of an enum would make it almost free to check.

Comment (jkotas) [reply src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/RuntimeInformation.Unix.cs:24]: These APIs did not get much use so far, and thus their performance did not really matter.

For .NET 5+, we are making these APIs a center point of larger end-to-end feature. It is why their performance matters for .NET 5+ more than before.

> higher initialization cost - we need to perform this check for every known OSPlatform (10 as of today, most probably will grow over the time)

Right, the design of this type is not particularly lean. It has non-trivial initialization cost designed in: We have to create 10 static fields, allocate 10 string literals. To cache the current platform check, we would also compare these string literals a few times. It does not make it that much more expensive.

> won't make much difference for scenarios where consumer library is using OSPlatform.Create

As long as the platform value created by OSPlatform.Create is cached, it will make the same difference. The name of the API `OSPlatform.Create` suggests that creating the platform moniker is not cheap. I do not feel bad about making `OSPlatform.Create` more expensive, assuming that people cache the result in a singleton.

> For all the use cases except checking for macOS on OSX

I can bet that the set of platforms that form equivalence or hierarchy will grow over time.

> the new APIs were supposed to be as fast as possible, they should rather be based on enum rather than string

The libraries design team has concerns about using enums for platform in public surface. The implementation can use enums internally, but we are stuck with strings in public surface. I do not think we would gain much by using enums vs. just `bool` internally. `bool` is sufficient to optimize the check that matters - whether the platform is the current platform.

Comment (adamsitnik) [reply src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/RuntimeInformation.Unix.cs:24]: @jkotas I've added the caching and also an xml doc remark to `.Create` method. Thank you!

File: src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/RuntimeInformation.Unix.cs
Comment (jkotas) [inline src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/RuntimeInformation.Unix.cs:24]:
```suggestion
            else if (name == "OSX")
```

File: src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/OSPlatform.cs
Comment (jkotas) [inline src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/OSPlatform.cs:31]: Comment about why we are caching `IsCurrent` may be useful

**Review by jkotas — APPROVED** [review]
Thanks!

**Review by jeffhandley — APPROVED** [review]

**Comment by jeffhandley** [comment]
Thanks, @adamsitnik and @jkotas!

**Pull request merged**
